### PR TITLE
IpAddress: bugfix and improvements

### PR DESCRIPTION
**Pull request by gx740**

### CODE REVIEW

File: Common++/src/IpAddress.cpp [hunk 1]
```diff
@@ -180,7 +178,7 @@ IPv6Address::IPv6Address(const IPv6Address& other)
 
 IPv6Address::~IPv6Address()
 {
-	delete m_pInAddr;
+	delete[] m_pInAddr;
```
Comment (seladb) [inline Common++/src/IpAddress.cpp]: why did you add `[]` here?

File: Common++/src/IpAddress.cpp [hunk 2]
```diff
@@ -246,10 +244,8 @@ bool IPv6Address::operator!=(const IPv6Address& other)
 
 IPv6Address& IPv6Address::operator=(const IPv6Address& other)
 {
-	if (m_pInAddr != NULL)
-		delete m_pInAddr;
-
-	m_pInAddr = (in6_addr*)new uint8_t[sizeof(in6_addr)];
+	if (m_pInAddr == NULL)
+		m_pInAddr = (in6_addr*)new uint8_t[sizeof(in6_addr)];
```
Comment (seladb) [inline Common++/src/IpAddress.cpp]: I'm not sure why you did this change. Could you please explain?

File: Common++/src/IpAddress.cpp [hunk 3]
```diff
@@ -131,10 +131,8 @@ uint32_t IPv4Address::toInt() const
 
 IPv4Address& IPv4Address::operator=(const IPv4Address& other)
 {
-	if (m_pInAddr != NULL)
-		delete m_pInAddr;
-
-	m_pInAddr = new in_addr();
+	if (m_pInAddr == NULL)
+		m_pInAddr = new in_addr();
```
Comment (seladb) [inline Common++/src/IpAddress.cpp]: I'm not sure why you did this change. Could you please explain?

File: Common++/src/IpAddress.cpp (hunk 1 above)
Comment (gx740) [reply Common++/src/IpAddress.cpp]: Because in constructor of IPv6Address the memory is allocated by by `new[]`. Refer to line 171.

`m_pInAddr = (in6_addr*)new uint8_t[sizeof(in6_addr)];`

The current implementation leads to memory leakage.

Comment (gx740) [reply Common++/src/IpAddress.cpp]: When I fixed this issue, the memory consumption decreased from 1.5Gbytes downto 490 Mbytes in my application.

Comment (seladb) [reply Common++/src/IpAddress.cpp]: Got it, that makes sense. Thanks for discovering this bug!

File: Common++/src/IpAddress.cpp (hunk 2 above)
Comment (gx740) [reply Common++/src/IpAddress.cpp]: The buffer `m_pInAddr` have the constant size `sizeof(in6_addr)`. The copy operation is made with the same size: `memcpy(m_pInAddr, other.m_pInAddr, sizeof(in6_addr))`. So if a pointer to buffer is not null we can just copy the data into it. It eliminates, at least, delete operator.
Besides, we can remove the condition `if (m_pInAddr == NULL)` because the buffer is always allocated by constructors.

File: Common++/src/IpAddress.cpp (hunk 3 above)
Comment (gx740) [reply Common++/src/IpAddress.cpp]: The same as for IPv6Address.

File: Common++/src/IpAddress.cpp (hunk 2 above)
Comment (gx740) [reply Common++/src/IpAddress.cpp]: More precisely, we can remove following code at all:
```
if (m_pInAddr == NULL)
  m_pInAddr = (in6_addr*)new uint8_t[sizeof(in6_addr)];

```
